README.md
# play-imagemagick
Play plugin that works with play-res to rescale images with the help of image magick

# Get PGP key
https://github.com/xerial/sbt-sonatype


# Release snapshot version
```bash
sbt publishSigned
```

# Release production version to maven central
remove -SNAPSHOT from version
```bash
sbt sonatypeRelease
```

CHANGELOG.md
# 0.1.0

 * Resize images using ImageMagick and IM4Java


test/se/digiplant/imagemagick/plugin/Spec.scala
package se.digiplant.imagemagick.plugin

import org.specs2.mutable.Around
import org.specs2.specification.Scope
import org.specs2.execute.{AsResult, Result}
import play.api.test._
import play.api.test.Helpers._
import java.io.File
import org.apache.commons.io.FileUtils
import util.Random

trait ScalrContext extends Around with TempFile {

  implicit val app: FakeApplication = new FakeApplication(
    additionalConfiguration = Map(
      "res.default" -> "tmp/default",
      "res.imagemagickcache" -> "tmp/imagemagickcache",
      "imagemagick.cache" -> "imagemagickcache",
      "imagemagick.cachedir" -> "tmp/imagemagickcachedir"
    )
  )

  def around[T : AsResult](t: =>T) = Helpers.running(app) {
    val result = AsResult.effectively(t)

    tmp.delete()

    result
  }
}

trait TempFile extends Scope {
  lazy val tmp = new File("tmp")
  lazy val logo = new File("test/resources/digiplant.jpg")
  lazy val LargeLogo = new File("test/resources/digiplant_large.jpg")

  def testFile: File = {
    tmp.mkdir()
    val chars = ('a' to 'z') ++ ('A' to 'Z') ++ ('1' to '9')
    val rand = (1 to 20).map(x => chars(Random.nextInt(chars.length))).mkString
    val tmpFile = new File("tmp", rand + ".jpg")
    FileUtils.copyFile(logo, tmpFile)
    tmpFile
  }
  def largeTestFile: File = {
    tmp.mkdir()
    val chars = ('a' to 'z') ++ ('A' to 'Z') ++ ('1' to '9')
    val rand = (1 to 20).map(x => chars(Random.nextInt(chars.length))).mkString
    val tmpFile = new File("tmp", rand + ".jpg")
    FileUtils.copyFile(LargeLogo, tmpFile)
    tmpFile
  }
}


test/se/digiplant/imagemagick/plugin/api/IMPluginSpec.scala
package se.digiplant.imagemagick.plugin.api

import org.specs2.mutable.Specification
import java.io.{File, FileInputStream}

import org.apache.commons.io.IOUtils
import javax.imageio.ImageIO

import org.im4java.core.IMOperation
import se.digiplant.imagemagick.plugin.ScalrContext
import se.digiplant.res.api.Res
import se.digiplant.scalr.api

object IMPluginSpec extends Specification {

  "IMPlugin Plugin" should {

    "resize image with default operations" in new ScalrContext {
      val fileuid = Res.put(testFile)
      fileuid.isEmpty must beFalse
      val file = new File("tmp/default/5564/ac5e/5564ac5e3968e77b4022f55a23d36630bdeb0274.jpg")
      file.exists() must beTrue

      val test = Res.get(fileuid).get

      val resizedSmall = IMPlugin.resize(test, IMPluginSettings.get("small_fit_to_width"))
      resizedSmall.exists() must beTrue

      val in1 = new FileInputStream(resizedSmall)
      val buf1 = ImageIO.read(in1)
      IOUtils.closeQuietly(in1)
      buf1.getWidth must beEqualTo(640)

      val resizedMedium = IMPlugin.resize(test, IMPluginSettings.get("medium_fit_to_height"))
      resizedMedium.exists() must beTrue

      val in2 = new FileInputStream(resizedMedium)
      val buf2 = ImageIO.read(in2)
      IOUtils.closeQuietly(in2)
      buf2.getHeight must beEqualTo(800)

      val resizedLarge = IMPlugin.resize(test, IMPluginSettings.get("large_fit_to_exact"))
      resizedLarge.exists() must beTrue

      val in3 = new FileInputStream(resizedLarge)
      val buf3 = ImageIO.read(in3)
      IOUtils.closeQuietly(in3)
      buf3.getWidth must beEqualTo(2560)
      buf3.getHeight must beEqualTo(1600)
    }

    "crop image with default operations" in new ScalrContext {
      val fileuid = Res.put(largeTestFile)
      fileuid.isEmpty must beFalse
      val file = new File("tmp/default/ebe5/b3ea/ebe5b3ea05c8faa0c3d1e71c0b113da8cac754a7.jpg")
      file.exists() must beTrue

      val test = Res.get(fileuid).get

      val croppedMedium = IMPlugin.resize(test, IMPluginSettings.get("medium_cropped"))
      croppedMedium.exists() must beTrue

      val in = new FileInputStream(croppedMedium)
      val buf = ImageIO.read(in)
      IOUtils.closeQuietly(in)
      buf.getWidth must beEqualTo(1280)
      buf.getHeight must beEqualTo(800)
    }

    "thumbnail image with default operations" in new ScalrContext {
      val fileuid = Res.put(testFile)
      fileuid.isEmpty must beFalse
      val file = new File("tmp/default/ebe5/b3ea/ebe5b3ea05c8faa0c3d1e71c0b113da8cac754a7.jpg")
      file.exists() must beTrue

      val test = Res.get(fileuid).get

      val thumbnailSmall = IMPlugin.resize(test, IMPluginSettings.get("small_thumbnail"))
      thumbnailSmall.exists() must beTrue

      val in = new FileInputStream(thumbnailSmall)
      val buf = ImageIO.read(in)
      IOUtils.closeQuietly(in)
      buf.getWidth must beEqualTo(50)
      buf.getHeight must beEqualTo(50)
    }

    "resize image with custom operations" in new ScalrContext {
      val fileuid = Res.put(testFile)
      fileuid.isEmpty must beFalse
      val file = new File("tmp/default/ebe5/b3ea/ebe5b3ea05c8faa0c3d1e71c0b113da8cac754a7.jpg")
      file.exists() must beTrue

      val test = Res.get(fileuid).get

      val op = new IMOperation()
      op.addImage()
      op.resize(500, 500);
      op.quality(90.0);
      op.addImage();

      IMPluginSettings.add("custom_small", op)
      val customSmall = IMPlugin.resize(test, IMPluginSettings.get("custom_small"))
      customSmall.exists() must beTrue

      val in1 = new FileInputStream(customSmall)
      val buf1 = ImageIO.read(in1)
      IOUtils.closeQuietly(in1)
      buf1.getWidth must beEqualTo(500)
      buf1.getHeight must beEqualTo(500)

      IMPluginSettings.add("custom_medium", IMPluginSettings.resize(1280, 800, '!', 80.0))
      val customMedium = IMPlugin.resize(test, IMPluginSettings.get("custom_medium"))
      customMedium.exists() must beTrue

      val in2 = new FileInputStream(customMedium)
      val buf2 = ImageIO.read(in2)
      IOUtils.closeQuietly(in2)
      buf2.getWidth must beEqualTo(1280)
      buf2.getHeight must beEqualTo(800)
    }

    "resize image with default setting" in new ScalrContext {
      val fileuid = Res.put(testFile)
      fileuid.isEmpty must beFalse
      val file = new File("tmp/default/ebe5/b3ea/ebe5b3ea05c8faa0c3d1e71c0b113da8cac754a7.jpg")
      file.exists() must beTrue

      val test = Res.get(fileuid).get

      val default1 = IMPlugin.resize(test, IMPluginSettings.get("default"))
      default1.exists() must beTrue

      val in1 = new FileInputStream(default1)
      val buf1 = ImageIO.read(in1)
      IOUtils.closeQuietly(in1)
      buf1.getWidth must beEqualTo(600)

      val default2 = IMPlugin.resize(test, IMPluginSettings.get("no_setting"))
      default2.exists() must beTrue

      val in2 = new FileInputStream(default2)
      val buf2 = ImageIO.read(in2)
      IOUtils.closeQuietly(in2)
      buf2.getWidth must beEqualTo(600)
    }

  }
}


test/se/digiplant/imagemagick/plugin/IMPluginAssetsSpec.scala
package se.digiplant.imagemagick.plugin

import org.specs2.mutable.Specification
import play.api.test._
import play.api.test.Helpers._

object IMPluginAssetsSpec extends Specification {

  "IMPlugin Controller" should {

    "return resized image in subfolder" in new ScalrContext {
      val result = IMPluginAssets.at("/test/resources", "subdir/test.jpg", "medium")(FakeRequest())
      status(result) must equalTo(OK)
      contentType(result) must beSome("image/jpeg")
    }

  }
}


sample/app/Global.scala
import scala.concurrent.ExecutionContext
import play.api._
import play.api.libs.json._
import play.modules.reactivemongo._
import play.modules.reactivemongo.json.collection.JSONCollection
import models._
import play.api.Play.current
import se.digiplant.imagemagick.plugin.IMPluginResAssets

object Global extends GlobalSettings with InitModels {

  override def onStart(app: Application) {
    initDB()
  }
}

trait InitModels {

  import ExecutionContext.Implicits.global

  def initDB() {
    val beachesFuture = Beach.collection.find(Json.obj()).cursor[JsObject].headOption
    beachesFuture.map { beach =>
      if (!beach.isDefined) {
        insertModels()
      }
    }
  }

  def insertModels() {
    // Create three beaches
    (1 to 3).map { i =>
      println(s"Saving beach $i")

      Beach.collection.insert(Beach(
        name = s"Beach $i",
        image = Image.fromPath(s"mockimages/beach$i.jpg", "default")
      ))
    }
  }
}


app/se/digiplant/imagemagick/plugin/IMPlugin.java
package se.digiplant.imagemagick.plugin;

import java.io.File;
import play.libs.*;

public class IMPlugin {

    /**
     * Creates and caches image in local cache directory
     * @param path The path to where the images are stored
     * @param file The filePath relative to the path variable (the same as the play Assets Controller)
     * @return a File if everything when well
     */
    public static File get(String path, String file, String setting) {
        return Scala.orNull(se.digiplant.imagemagick.plugin.api.IMPlugin.get(path, file, setting));
    }

    public static File get(String path, String file) {
        return get(path, file, "default");
    }

    /**
     *
     * @param fileuid The unique play-res file identifier
     * @param source The play-res source name
     * @return a File if everything when well
     */
    public static File getRes(String fileuid, String source, String setting) {
        return Scala.orNull(se.digiplant.imagemagick.plugin.api.IMPlugin.getRes(fileuid, source, setting));
    }

    public static File getRes(String fileuid, String source) {
        return getRes(fileuid, source, "default");
    }
}


app/se/digiplant/imagemagick/plugin/api/IMPlugin.scala
package se.digiplant.imagemagick.plugin.api

import play.api._
import java.io.{File, FileInputStream, OutputStream}

import org.apache.commons.io.{FilenameUtils, IOUtils}

import org.im4java.core.{ConvertCmd, IMOperation}

import util.Random
import se.digiplant.res.api.Res
import play.api.Play.current

object IMPlugin {

  lazy val configuration = Play.configuration.getConfig("imagemagick").getOrElse(Configuration.empty)

  /**
   * Creates and caches image in local cache directory
    *
    * @param path The path to where the images are stored
   * @param file The filePath relative to the path variable (the same as the play Assets Controller)
   * @return a File if everything when well
   */
  def get(path: String, file: String, setting: String): Option[File] = {

    val resourceName = Option(path + "/" + file).map(name => if (name.startsWith("/")) name else ("/" + name)).get
    val resourceFile = Play.getFile(resourceName)

    if (resourceFile.isDirectory) {
      None
    } else {
      val cachePath = configuration.getString("cachedir").getOrElse("tmp/imagemagickcache")
      val cachedImage = Res.fileWithMeta(
        filePath = FilenameUtils.concat(cachePath, file),
        meta = Seq(setting)
      )

      cachedImage.map { cachedImage =>
        Some(cachedImage)
      }.getOrElse {
        val resizedImage = resize(resourceFile, IMPluginSettings.get(setting))
        val resizedFilePath = Res.saveWithMeta(
          resizedImage,
          filePath = FilenameUtils.concat(cachePath, file),
          meta = Seq(setting)
        )
        Play.getExistingFile(resizedFilePath)
      }
    }
  }

  /**
   *
   * @param fileuid The unique play-res file identifier
   * @param source The play-res source name
   * @return a File if everything when well
   */
  def getRes(fileuid: String, source: String = "default", setting: String = "default"): Option[File] = {

    Res.get(fileuid, source).flatMap { res =>

      val cacheSource = configuration.getString("cache").getOrElse("imagemagickcache")

      val cachedImage = Res.get(
        fileuid = res.getName,
        meta = Seq(setting),
        source = cacheSource
      )

      cachedImage.map { cachedImage =>
        Some(cachedImage)
      }.getOrElse {
        val resizedImage = resize(res, IMPluginSettings.get(setting))
        val fileUID = Res.put(
          resizedImage,
          cacheSource,
          filename = Some(res.getName),
          meta = Seq(setting)
        )
        Res.get(fileUID, cacheSource)
      }
    }
  }

  /**
   * Resizes an image and stores it in a temp file that we then can move into a cache
   *
   * @param file The file to resize
   * @return A resized file
   */
  def resize(file: File, op: IMOperation): File = {
    // create temp file
    val ext = guessImageFormat(file)
    val tmp = File.createTempFile(Random.nextString(20), ext)

    // create convert command
    val cmd = new ConvertCmd()

    // execute the operation
    cmd.run(op, file.getPath, tmp.getPath);
    tmp
  }

  /**
   * Detect the image format "jpg" or "png"
   * Done by reading the first bytes of the file to check if it correspond to a file format "signature"
    *
    * @see http://www.garykessler.net/library/file_sigs.html
   * @param file The File we want to analyse
   * @return A String representing the image format "jpg" or "png"
   */
  def guessImageFormat(file: File): String = {
    val is = new FileInputStream(file)
    val firstBytes = new Array[Byte](2)
    is.read(firstBytes)
    is.close
    // A JPEG file always start with bytes FF D8
    if (firstBytes(0) == 0xFF.asInstanceOf[Byte] && firstBytes(1) == 0xD8.asInstanceOf[Byte])
      "jpg"
    else
      "png"
  }
}


app/se/digiplant/imagemagick/plugin/IMPluginAssets.scala
package se.digiplant.imagemagick.plugin

import play.api._
import play.api.mvc._
import play.api.libs._
import org.joda.time.format.{DateTimeFormat, DateTimeFormatter}
import org.joda.time.DateTimeZone

import collection.JavaConverters._
import java.io.File

import org.im4java.core.IMOperation
import play.api.Play.current

object IMPluginAssets extends Controller {

  private val timeZoneCode = "GMT"

  //Dateformatter is immutable and threadsafe
  private val df: DateTimeFormatter = DateTimeFormat
    .forPattern("EEE, dd MMM yyyy HH:mm:ss '" + timeZoneCode + "'")
    .withLocale(java.util.Locale.ENGLISH)
    .withZone(DateTimeZone.forID(timeZoneCode))

  //Dateformatter is immutable and threadsafe
  private val dfp: DateTimeFormatter = DateTimeFormat
    .forPattern("EEE, dd MMM yyyy HH:mm:ss")
    .withLocale(java.util.Locale.ENGLISH)
    .withZone(DateTimeZone.forID(timeZoneCode))

  private val parsableTimezoneCode = " " + timeZoneCode

  /**
   * Resize and cache images stored in play-res
   * @param path Path relative to play app
   * @param file Filepath of the file
   * @return A resized image
   */
  def at(path: String, file: String, setting: String): Action[AnyContent] = Action { request =>


    def parseDate(date: String): Option[java.util.Date] = {
      try {
        //jodatime does not parse timezones, so we handle that manually
        val d = dfp.parseDateTime(date.replace(parsableTimezoneCode, "")).toDate
        Some(d)
      } catch {
        case _: Exception => None
      }
    }

    api.IMPlugin.get(path, file, setting).map { resizedImage =>
      request.headers.get(IF_NONE_MATCH).flatMap {
        ifNoneMatch => etagFor(resizedImage).filter(_ == ifNoneMatch)
      }.map(_ => NotModified).getOrElse {
        request.headers.get(IF_MODIFIED_SINCE).flatMap(parseDate).flatMap {
          ifModifiedSince => lastModifiedFor(resizedImage).flatMap(parseDate).filterNot(lastModified => lastModified.after(ifModifiedSince))
        }.map(_ => NotModified.withHeaders(
          DATE -> df.print({ new java.util.Date }.getTime)
        )).getOrElse {

          val response = Ok.sendFile(resizedImage, inline = true)

          // Add Etag if we are able to compute it
          val taggedResponse = etagFor(resizedImage).map(etag => response.withHeaders(ETAG -> etag)).getOrElse(response)
          val lastModifiedResponse = lastModifiedFor(resizedImage).map(lastModified => taggedResponse.withHeaders(LAST_MODIFIED -> lastModified)).getOrElse(taggedResponse)

          // Add Cache directive if configured

          /*val cachedResponse = lastModifiedResponse.withHeaders(CACHE_CONTROL -> {
            Play.mode match {
              case Mode.Prod => Play.configuration.getString("assets.defaultCache").getOrElse("max-age=3600")
              case _ => "no-cache"
            })
          })*/

          lastModifiedResponse

        }: Result
      }
    }.getOrElse {
      NotFound
    }
  }

  // Last modified
  private val lastModifieds = (new java.util.concurrent.ConcurrentHashMap[String, String]()).asScala

  private def lastModifiedFor(file: File): Option[String] = {
    lastModifieds.get(file.getName).filter(_ => Play.isProd).orElse {
      val lastModified = df.print({
        new java.util.Date(file.lastModified).getTime
      })
      lastModifieds.put(file.getName, lastModified)
      Some(lastModified)
    }
  }

  // Etags
  private val etags = (new java.util.concurrent.ConcurrentHashMap[String, String]()).asScala

  private def etagFor(file: File): Option[String] = {
    etags.get(file.getName).filter(_ => Play.isProd).orElse {
      val maybeEtag = lastModifiedFor(file).map(_ + " -> " + file.getName).map("\"" + Codecs.sha1(_) + "\"")
      maybeEtag.foreach(etags.put(file.getName, _))
      maybeEtag
    }
  }

}

app/se/digiplant/imagemagick/plugin/IMPluginResAssets.scala
package se.digiplant.imagemagick.plugin

import play.api._
import play.api.mvc._
import play.api.libs._
import org.joda.time.format.{DateTimeFormat, DateTimeFormatter}
import org.joda.time.DateTimeZone

import collection.JavaConverters._
import java.io.File

import org.im4java.core.IMOperation
import play.api.Play.current

object IMPluginResAssets extends Controller {

  private val timeZoneCode = "GMT"

  //Dateformatter is immutable and threadsafe
  private val df: DateTimeFormatter = DateTimeFormat
    .forPattern("EEE, dd MMM yyyy HH:mm:ss '" + timeZoneCode + "'")
    .withLocale(java.util.Locale.ENGLISH)
    .withZone(DateTimeZone.forID(timeZoneCode))

  //Dateformatter is immutable and threadsafe
  private val dfp: DateTimeFormatter = DateTimeFormat
    .forPattern("EEE, dd MMM yyyy HH:mm:ss")
    .withLocale(java.util.Locale.ENGLISH)
    .withZone(DateTimeZone.forID(timeZoneCode))

  private val parsableTimezoneCode = " " + timeZoneCode

  /**
   * Resize and cache images stored in play-res
   * @param fileuid Fileuid of file stored in play-res
   * @param source play-res source
   * @return A resized image
   */
  def at(fileuid: String, setting: String = "default", source: String = "default"): Action[AnyContent] = Action { request =>

    def parseDate(date: String): Option[java.util.Date] = {
      try {
        //jodatime does not parse timezones, so we handle that manually
        val d = dfp.parseDateTime(date.replace(parsableTimezoneCode, "")).toDate
        Some(d)
      } catch {
        case _: Exception => None
      }
    }

    api.IMPlugin.getRes(fileuid, source, setting).map { resizedImage =>
      request.headers.get(IF_NONE_MATCH).flatMap {
        ifNoneMatch => etagFor(resizedImage).filter(_ == ifNoneMatch)
      }.map(_ => NotModified).getOrElse {
        request.headers.get(IF_MODIFIED_SINCE).flatMap(parseDate).flatMap {
          ifModifiedSince => lastModifiedFor(resizedImage).flatMap(parseDate).filterNot(lastModified => lastModified.after(ifModifiedSince))
        }.map(_ => NotModified.withHeaders(
          DATE -> df.print({ new java.util.Date }.getTime)
        )).getOrElse {

          val response = Ok.sendFile(resizedImage, inline = true)

          // Add Etag if we are able to compute it
          val taggedResponse = etagFor(resizedImage).map(etag => response.withHeaders(ETAG -> etag)).getOrElse(response)
          val lastModifiedResponse = lastModifiedFor(resizedImage).map(lastModified => taggedResponse.withHeaders(LAST_MODIFIED -> lastModified)).getOrElse(taggedResponse)

          // Add Cache directive if configured

          /*val cachedResponse = lastModifiedResponse.withHeaders(CACHE_CONTROL -> {
            Play.mode match {
              case Mode.Prod => Play.configuration.getString("assets.defaultCache").getOrElse("max-age=3600")
              case _ => "no-cache"
            })
          })*/

          lastModifiedResponse

        }: Result
      }
    }.getOrElse {
      NotFound
    }
  }

  // Last modified
  private val lastModifieds = (new java.util.concurrent.ConcurrentHashMap[String, String]()).asScala

  private def lastModifiedFor(file: File): Option[String] = {
    lastModifieds.get(file.getName).filter(_ => Play.isProd).orElse {
      val lastModified = df.print({
        new java.util.Date(file.lastModified).getTime
      })
      lastModifieds.put(file.getName, lastModified)
      Some(lastModified)
    }
  }

  // Etags
  private val etags = (new java.util.concurrent.ConcurrentHashMap[String, String]()).asScala

  private def etagFor(file: File): Option[String] = {
    etags.get(file.getName).filter(_ => Play.isProd).orElse {
      val maybeEtag = lastModifiedFor(file).map(_ + " -> " + file.getName).map("\"" + Codecs.sha1(_) + "\"")
      maybeEtag.foreach(etags.put(file.getName, _))
      maybeEtag
    }
  }

}


app/se/digiplant/imagemagick/plugin/api/IMPluginSettings.java
package se.digiplant.imagemagick.plugin.api;

import org.im4java.core.IMOperation;

import java.util.HashMap;

public class IMPluginSettings {

    public static HashMap<String, IMOperation> settings = new HashMap<>();

    public static void add(String key, IMOperation op) {
        settings.put(key, op);
    }

    public static IMOperation get(String setting) {
        if(settings.containsKey(setting)) {
            return settings.get(setting);
        }

        return getDefault(setting);
    }

    public static IMOperation getDefault(String key) {
        switch (key) {
            case "small": return resize(600, 400, null, 100.0);
            case "medium": return resize(1280, 800, null, 100.0);
            case "large": return resize(2560, 1600, null, 100.0);

            case "small_optimized": return resize(640, 480, null, 80.0);
            case "medium_optimized": return resize(1280, 800, null, 80.0);
            case "large_optimized": return resize(2560, 1600, null, 80.0);

            case "small_fit_to_width": return resize(640, null, null, 100.0);
            case "medium_fit_to_width": return resize(1280, null, null, 100.0);
            case "large_fit_to_width": return resize(2560, null, null, 100.0);

            case "small_fit_to_height": return resize(null, 480, null, 100.0);
            case "medium_fit_to_height": return resize(null, 800, null, 100.0);
            case "large_fit_to_height": return resize(null, 1600, null, 100.0);

            case "small_fit_exact": return resize(640, 400, '!', 100.0);
            case "medium_fit_to_exact": return resize(1280, 800, '!', 100.0);
            case "large_fit_to_exact": return resize(2560, 1600, '!', 100.0);

            case "small_cropped": return crop(640, 400, 0, 0, 100.0);
            case "medium_cropped": return crop(1280, 800, 0, 0, 100.0);
            case "large_cropped": return crop(2560, 1600, 0, 0, 100.0);

            case "small_cropped_optimized": return crop(640, 400, 0, 0, 80.0);
            case "medium_cropped_optimized": return crop(1280, 800, 0, 0, 80.0);
            case "large_cropped_optimized": return crop(2560, 1600, 0, 0, 80.0);

            case "small_thumbnail": return thumbnail(50, 50, 100.0);
            case "medium_thumbnail": return thumbnail(100, 100, 100.0);
            case "large_thumbnail": return thumbnail(150, 150, 100.0);

            case "small_thumbnail_optimized": return thumbnail(50, 50, 80.0);
            case "medium_thumbnail_optimized": return thumbnail(100, 100, 80.0);
            case "large_thumbnail_optimized": return thumbnail(150, 150, 80.0);

            default: return resize(800, 600, null, 100.0);
        }
    }

    public static IMOperation resize(Integer width, Integer height, Character flag, double quality) {
        IMOperation op = new IMOperation();

        op.addImage();
        op.resize(width, height, flag);
        op.quality(quality);
        op.addImage();

        return op;
    }

    public static IMOperation crop(Integer width, Integer height, Integer offsetX, Integer offsetY, double quality) {
        IMOperation op = new IMOperation();

        op.addImage();
        op.crop(width, height, offsetX, offsetY);
        op.quality(quality);
        op.addImage();

        return op;
    }

    public static IMOperation thumbnail(Integer width, Integer height, double quality) {
        IMOperation op = new IMOperation();

        op.addImage();
        op.thumbnail(width, height);
        op.quality(quality);
        op.addImage();

        return op;
    }


}
